src/network/networkd-netlabel.c

```diff
@@ -0,0 +1,197 @@
+                               address->link);
+        if (r < 0)
+                return r;
+
+        link_ref(address->link);
+        return 0;
+}
+
+static const char *get_label(const Address *address) {
+        assert(address);
+        assert(address->link);
+
+        if (!address->link->network || !IN_SET(address->family, AF_INET, AF_INET6))
+                return NULL;
+
+        switch (address->source) {
+        case NETWORK_CONFIG_SOURCE_DHCP4:
+                return address->link->network->dhcp_netlabel;
+        case NETWORK_CONFIG_SOURCE_DHCP6:
+                return address->link->network->dhcp6_netlabel;
+        case NETWORK_CONFIG_SOURCE_DHCP_PD:
+                return address->link->network->dhcp_pd_netlabel;
+        case NETWORK_CONFIG_SOURCE_NDISC:
+                return address->link->network->ndisc_netlabel;
+        case NETWORK_CONFIG_SOURCE_STATIC:
+                return address->netlabel;
+        default:
+                return NULL;
+        }
+}
+
+void address_add_netlabel(const Address *address) {
+        int r;
+        assert(address);
+
+        const char *label;
+        label = get_label(address);
+        if (!label)
+                return;
+
+        r = netlabel_command(NLBL_UNLABEL_C_STATICADD, label, address);
+        if (r < 0)
+                log_link_warning_errno(address->link, r, "Adding NetLabel %s for IP address %s failed, ignoring", label,
+                                       IN_ADDR_PREFIX_TO_STRING(address->family, &address->in_addr, address->prefixlen));
+        else
+                log_link_debug(address->link, "Adding NetLabel %s for IP address %s", label,
+                               IN_ADDR_PREFIX_TO_STRING(address->family, &address->in_addr, address->prefixlen));
+}
+
+void address_del_netlabel(const Address *address) {
+        int r;
+        assert(address);
+
+        const char *label;
+        label = get_label(address);
+        if (!label)
+                return;
+
+        r = netlabel_command(NLBL_UNLABEL_C_STATICREMOVE, label, address);
+        if (r < 0)
+                log_link_warning_errno(address->link, r, "Deleting NetLabel %s for IP address %s failed, ignoring", label,
+                                       IN_ADDR_PREFIX_TO_STRING(address->family, &address->in_addr, address->prefixlen));
+        else
+                log_link_debug(address->link, "Deleting NetLabel %s for IP address %s", label,
+                               IN_ADDR_PREFIX_TO_STRING(address->family, &address->in_addr, address->prefixlen));
+}
+
+int config_parse_netlabel(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+        char **label = ASSERT_PTR(data);
```

Comment:
Please insert an empty line before variable declaration and function arguments.

Reply:
inserted